src/rgw/rgw_auth_s3.cc

```diff
@@ -508,15 +508,18 @@ static inline std::string aws4_uri_recode(const boost::string_view& src)
 
 std::string get_v4_canonical_qs(const req_info& info, const bool using_qs)
 {
-  if (info.request_params.empty()) {
+  std::string request_params = info.request_params;
+  if (request_params.empty()) {
     /* Optimize the typical flow. */
     return std::string();
+  } else {
+    boost::replace_all(request_params, "+", "%20");
```

Comment:
In [the AWSv4 documentation](http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) we can find:

> Percent-encode all other characters with `%XY`, where `X` and `Y` are hexadecimal characters (`0-9` and uppercase `A-F`). For example, the space character must be encoded as `%20` (not using `'+'`, as some encoding schemes do) and extended UTF-8 characters must be in the form `%XY%ZA%BC`.